app/grandchallenge/serving/models.py

```diff
@@ -43,3 +45,36 @@ class Download(models.Model):
     algorithm_image = models.ForeignKey(
         AlgorithmImage, null=True, on_delete=models.CASCADE, editable=False
     )
+
+
+def get_component_interface_values_for_user(*, user):
+    job_inputs_query = get_objects_for_user(
+        user=user, perms="algorithms.view_job"
+    ).filter(inputs__pk__in=OuterRef("pk"))
+
+    job_outputs_query = get_objects_for_user(
+        user=user, perms="algorithms.view_job"
+    ).filter(outputs__pk__in=OuterRef("pk"))
+
+    display_set_query = get_objects_for_user(
+        user=user, perms="reader_studies.view_displayset"
+    ).filter(values__pk__in=OuterRef("pk"))
+
+    archive_item_query = get_objects_for_user(
+        user=user, perms="archives.view_archiveitem"
+    ).filter(values__pk__in=OuterRef("pk"))
+
+    return (
+        ComponentInterfaceValue.objects.annotate(
+            has_view_job_inputs_perm=Exists(job_inputs_query)
+        )
+        .annotate(has_view_job_outputs_perm=Exists(job_outputs_query))
+        .annotate(has_view_ds_perm=Exists(display_set_query))
+        .annotate(has_view_ai_perm=Exists(archive_item_query))
```

Comment:
For my understanding: is there a reason for separate annotation clauses? Do we need to consider things like https://docs.djangoproject.com/en/5.1/topics/db/aggregation/#combining-multiple-aggregations ? We have to make sure this bit of code is rock solid. Are there comprehensive tests to cover the cases mentioned in the Django documentation?

Reply:
I have done some testing in production. The new method is just as performant as the old one, so that's great! It is still a little slow. If you put everything into one aggregation block rather than 4 it is consistently faster, so would be good to know the motivation for using separate ones.

I checked with 4 users: the anonymous user, a superuser, a user with many permissions who does not have access, a user who has access.

Reply:
I did separate annotations because we once had issues with combining multiple annotations (see the link you posted). 
But I think we might be able to simplify it by doing:

```
ComponentInterfaceValue.objects.annotate(
            has_permissions=Exists(job_inputs_query)
                            | Exists(job_outputs_query)
                            | Exists(display_set_query)
                            | Exists(archive_item_query)
        )
        .filter(has_permissions=True)
```

Need to test this out to make sure this works fully though.

Reply:
That is cleaner, maybe be a bit more specific on the name of the annotation (has_view_permission?) performance wise there's no difference and returns the correct result with the 4 user test I did.

Reply:
Thanks for checking! I have updated the method and also did some tests and can confirm that it returns the correct files. How slow is it in production, James?

Reply:
~50 ms for normal users, 1.6s for the user with lots of perms which is not great but at least not a regression.